README.qmd

```diff
@@ -446,9 +452,814 @@ rm(x)
+<br><br>
+
+### Exercises
+
+*1. How do you create raw and complex scalars? (See `?raw` and `?complex`.)*
+
+Raw vectors are created with `raw()`, specifying the single `length` argument. Raw vectors also have a specific `is.raw()` for checking and `as.raw()` for coercion.
+```{r}
+r <- raw(2)
+r
+```
+<br>
+
+Complex vectors are created with `complex()` specifying either the length (via the `length.out` argument), or both the real and imaginary parts as numeric vectors.
+<br><br>
+
+*2. Test your knowledge of the vector coercion rules by predicting the output of the following uses of `c()`*
+
+Per the type coercion hierarchy, `c(1, FALSE)` will be converted to a double, `c(1, 0)`. 1's are coerced to `TRUE` (and 0 to `FALSE`), and vice versa
+```{r}
+x <- c(1, FALSE)
+cat(x, " | ", typeof(x))
+```
+<br>
+
+`c("a", 1)`, will be coerced to `c("a", "1")` as the character type is about double in the type coversion hierarchy
+```{r}
+x <- c("a", 1)
+cat(x, " | ", typeof(x))
+```
+<br>
+
+`c(TRUE, 1L)`, will b coerced to `c(1L, 1L)` per the type coercion hierarchy
+```{r}
+x <- c(TRUE, 1L)
+cat(x, " | ", typeof(x))
+```
+<br>
+
+*3.1. Why is `1 == "1"` true?*
+
+Per the documentation for `==`, atomic vectors that are of different types are coerced prior to evaluation. So, the left hand of the equality, `1`, will be coerced to a character vector prior to evaluation. Once coerced, the call is `"1" == "1"`, which is clearly true.
+
+*3.2. Why is `-1 < FALSE` true?*
+
+`<`, like `==` and most operations, also coerces types prior to evaluation. So after coercion, the call here becomes `-1 < 0`, which is false.
+
+*3.3. Why is `"one" < 2` false?*
+
+Per the documentation for `<`, string comparison is done at the locale level. The locale in use can be viewed with `Sys.getlocale()` and similarly set with `Sys.setlocale()`. Because numbers come before letters in this sequence, the coerced inequality, `"one" < "2"` is evaluated as `FALSE`.
+<br><br>
+
+*4. Why is the default missing value, `NA`, a logical vector? What’s special about logical vectors? (Hint: think about `c(FALSE, NA_character_)`.)*
+
+Logical vectors are lowest on the type hierarchy. When they are combined with another primary type, logicals will always be coerced into the other type.
+<br><br>
+
+*5. Precisely what do `is.atomic()`, `is.numeric()`, and `is.vector()` test for?*
+
+- `is.atomic()` checks if the vector is of type "logical", "integer", "numeric", "complex", "character" or "raw"
+- `is.numeric()` checks if the vector is a double or integer and *not* a factor.
+- `is.vector()` is a generalized `is.*()` function for all vectors. The `mode` argument can be specified to check for a specific type, or it can be left as `"any"`, the default, to check is the object is a vector. `mode` can also be specified as `"numeric"`, running the same check as `is.numeric()`.
```

Comment:
```suggestion
- `is.vector()` is a generalized `is.*()` function for all vectors. The `mode` argument can be specified to check for a specific type, including lists. It can also be left as "any", the default, to check is the object is a vector. `mode` can also be specified as "numeric", running the same check as `is.numeric()`.
```